[IR TX RX] Extended AC via JSON formated commands

jimmys01 opened the pull request:
Example:
IRSENDAC,"protocol":"GREE","power":"off","mode":"heat","temp":21,"fanspeed":"max","swingv":"middle","swingh":"off"}

jimmys01 committed:
[IR TX RX] Library updates - New Protocols

jimmys01 committed:
Add the new IR protocols

Also added back the ability to have arbitrary bits to the IR send command. Reflected that also in the received signal decoding. Using 0 bits as an argument the ir send command will still use the default protocol bits
ref: https://www.letscontrolit.com/forum/viewtopic.php?f=4&t=6452&p=37373#p37373

jimmys01 committed:
[JSON] Update to Json 6

jimmys01 committed:
Update code to version JSON 6

jimmys01 committed:
Merge pull request #1 from jimmys01/JSON

Json

jimmys01 committed:
[IR TX RX] Extended AC commands

Example:
IRSENDAC,{"Protocol":"GREE","Power":true,"Opmode":"fan_only","Degrees":27,"Fanspeed":"min","Swingv":"middle","Swingh":"off","Light":true}

jimmys01 commented:
This PR also has the JSON and IR library updates built in.

jimmys01 committed:
Move defines in plugins

TD-er commented:
Travis errors:
```
/home/travis/build/letscontrolit/ESPEasy/src/_P016_IR.ino: In function 'boolean Plugin_016(byte, EventStruct*, String&)':
/home/travis/build/letscontrolit/ESPEasy/src/_P016_IR.ino:457:9: error: '*((void*)& state +38)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
if (state.beep)
^
/home/travis/build/letscontrolit/ESPEasy/src/_P016_IR.ino:455:9: error: '*((void*)& state +37)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
if (state.clean)
^
/home/travis/build/letscontrolit/ESPEasy/src/_P016_IR.ino:453:9: error: '*((void*)& state +36)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
if (state.filter)
^
/home/travis/build/letscontrolit/ESPEasy/src/_P016_IR.ino:451:9: error: '*((void*)& state +35)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
if (state.light)
^
/home/travis/build/letscontrolit/ESPEasy/src/_P016_IR.ino:449:9: error: '*((void*)& state +34)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
if (state.econo)
^
/home/travis/build/letscontrolit/ESPEasy/src/_P016_IR.ino:447:9: error: '*((void*)& state +33)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
if (state.turbo)
^
/home/travis/build/letscontrolit/ESPEasy/src/_P016_IR.ino:445:9: error: '*((void*)& state +32)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
if (state.quiet)
^
/home/travis/build/letscontrolit/ESPEasy/src/_P016_IR.ino:444:28: error: '*((void*)& state +28)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
doc[F("Swingh")] = IRac::swinghToString(state.swingh); //Horizontal Swing setting
^
/home/travis/build/letscontrolit/ESPEasy/src/_P016_IR.ino:442:28: error: '*((void*)& state +24)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
doc[F("Swingv")] = IRac::swingvToString(state.swingv); //Vertical swing setting
^
/home/travis/build/letscontrolit/ESPEasy/src/_P016_IR.ino:440:30: error: '*((void*)& state +20)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
doc[F("Fanspeed")] = IRac::fanspeedToString(state.fanspeed); //Fan Speed setting
^
/home/travis/build/letscontrolit/ESPEasy/src/_P016_IR.ino:437:20: error: '*((void*)& state +16)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
if (!state.celsius)
^
cc1plus: some warnings being treated as errors
*** [.pioenvs/normal_core_252_IR_ESP8266_4M/src/ESPEasy.ino.cpp.o] Error 1
```

jimmys01 committed:
Make travis happy and some code refactoring.

jimmys01 commented:
If anybody want to give it a try: [firmware.zip](https://github.com/letscontrolit/ESPEasy/files/3280373/firmware.zip)

jimmys01 commented:
> Linking .pioenvs/normal_IR_ESP8266_1M/firmware.elf
> /home/travis/.platformio/packages/toolchain-xtensa@1.40802.0/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/normal_IR_ESP8266_1M/firmware.elf section `.irom0.text' will not fit in region `irom0_0_seg'
> collect2: error: ld returned 1 exit status
> *** [.pioenvs/normal_IR_ESP8266_1M/firmware.elf] Error 1

Looks like 1M builds with extended AC commands is a no go
Maybe we can disable plugins that typicaly wont be usefull with IR builds?
edit: or 1M builds with `P016_P035_Extended_AC `disabled

jimmys01 committed:
Some defines fixes.

this wont fix the 1M build.

jimmys01 commented:
@TD-er now its up to you, I propose to just disable the P016_P035_Extended_AC  for the 1M build

> Memory Usage -> http://bit.ly/pio-memory-usage
> DATA:    [======    ]  60.3% (used 49396 bytes from 81920 bytes)
> PROGRAM: [==========]  98.4% (used 878916 bytes from 892912 bytes)

this will leave 13K available in the 1M ir build

jimmys01 committed:
Bug fixes

jimmys01 committed:
Some usage instructions added

jimmys01 commented:
If anybody want to give it a try: 
[firmware.zip](https://github.com/letscontrolit/ESPEasy/files/3293112/firmware.zip)

jimmys01 committed:
Update IR Lib

jimmys01 committed:
Exclude Extended AC commands from the 1M build

jimmys01 commented:
Where did it exactly failed? I cannot see it throwing an error at all.

EDIT: 
> The job exceeded the maximum time limit for jobs, and has been terminated

jimmys01 committed:
Force travis rebuild

Hopefully this time travis wont time out

TD-er commented:
We should have a good look of all the builds we now have and also what combinations are really needed for Travis builds and which one for nightly builds.

jimmys01 commented:
Did anybody else besides me  tested this PR? @clumsy-stefan  do you still test the JSON update?

clumsy-stefan commented:
@jimmys01 I still use #2489 in my builds and they are working fine for me...  not this one though..

jimmys01 commented:
I also use this PR in production, no problems at all so far...

jimmys01 committed:
Even more Ir Library updates

jimmys01 committed:
Add the new IR protocol

TD-er commented:
Any suggestion on how to proceed with the PR's?
First #2489 and then this one?

clumsy-stefan commented:
@TD-er I think also #2489 should be merged first (JSON only) and based on this further updates. Together with the new esp8266 core this renders a really stable version (for me at least)...

jimmys01 committed:
Follow the IR library JSON format

Since markzabo also implemented JSON, I thought it is better to have the same JSON format, since its the same library after all.

TD-er commented:
So I should also make core 2.5.2 the default?
Maybe also good to split core versions into separate ZIP files for download? The list of bin files is a bit long right now.

jimmys01 commented:
If you go ahead and merge 2489 (JSON) first so I can also refactor the JSON for the zabbix controler.

jimmys01 commented:
Don't merge this one yet, there is a juicy lirary update that will reduce the build by 3K, mayde then it will fit in the 1M builds to

jimmys01 committed:
Update IR Library

jimmys01 committed:
Re-enable Extended AC for 1M build

This version saves just enought space so that Exetnded AC commands fit in 1M build too.

jimmys01 commented:
Ready to commit

happytm commented:
Jimmy,

I am trying to compile your IR plugins along with some other plugins in arduino but without success.If you are able to compile can you please compile and upload bin file with following setup :

_C005
_C008
_C0010
_C0011
_C0013
_C0016


_N001
_N002


_P001
_P002
_P003
_P007
_P011
_P015
_P016
_P023
_P026
_P028
_P029
_P030
_P033
_P035
_P036
_P037
_P038
_P045
_P048
_P062
_P064
_P079

This will be for wemos d1 mini 4M version.

Also which one of above plugins would not work on ESP32 boards ?

I appreciate your help.

Thanks

jimmys01 commented:
> Also which one of above plugins would not work on ESP32 boards ?
That I do not know.

I recomend to use paltformio (I use it with visual studio code) to do this work. If you use Arduino IDE it can be done but it get messy with the library versioning. 

[firmware.zip](https://github.com/letscontrolit/ESPEasy/files/3353447/firmware.zip)

happytm commented:
Thank you so much. Appreciate it.

Have you tried  any of following sensors  on ESP32?

Your IR Plugins
BME280 Environment
APDS-9960 Gesture
SSD1306 Oled display

If so do they work?

Thanks again.

TD-er commented:
The BME280 and OLED have been known to work on ESP32.
I have used both in the past to test I2C issues on ESP32.

happytm commented:
Thanks TD-er.

Jimmy, your bin file worked fine. I had no chance to check until now. thank you again.

I appreciate both.

Only problem is it worked fine in dynamic ip mode but as soon as I entered fixed ip in gui it started restarting in loop with following message:

wdt reset
load 0x4010f000, len 1384, room 16 
tail 8
chksum 0x2d
csum 0x2d
vbb28d4a3
~ld
e⸮U64 : 


INIT : Booting version:  (ESP82xx Core 2_4_2, NONOS SDK 2.2.1(cfd48f3), LWIP: 2.0.3 PUYA support)
65 : INIT : Free RAM:27408
65 : INIT : Warm boot #54 - Restart Reason: Hardware Watchdog
68 : FS   : Mounting...
93 : FS   : Mount successful, used 75802 bytes of 957314
101 : CRC  : No program memory checksum found. Check output of crc2.py
111 : CRC  : SecuritySettings CRC   ...OK 
219 : INIT : Free RAM:24576
221 : INIT : I2C
221 : INIT : SPI not enabled
225 : INFO : Plugins: 22 [Normal] [IR_Extended] (ESP82xx Core 2_4_2, NONOS SDK 2.2.1(cfd48f3), LWIP: 2.0.3 PUYA support)
228 : WIFI : Set WiFi to STA
261 : WIFI : Connecting HTM1 attempt #0
262 : IP   : Static IP : 10.1.10.70 GW: 10.1.10.1 SN: 255.255.255.0 DNS: 10.1.10.1
264 : WIFI : Not configured in Station Mode!!: HTM1
1601 : WD   : Uptime 0 ConnectFailures 0 FreeMem 23216 WiFiStatus 0
4146 : WIFI : Connected! AP: HTM1 (00:27:22:E2:F0:52) Ch: 7 Duration: 3781 ms
4148 : IP   : Static IP : 10.1.10.70 GW: 10.1.10.1 SN: 255.255.255.0 DNS: 10.1.1
 ets Jan  8 2013,rst cause:4, boot mode:(3,7)

wdt reset
load 0x4010f000, len 1384, room 16 
tail 8
chksum 0x2d
csum 0x2d
vbb28d4a3
~ld
e⸮U74 : 


INIT : Booting version:  (ESP82xx Core 2_4_2, NONOS SDK 2.2.1(cfd48f3), LWIP: 2.0.3 PUYA support)
75 : INIT : Free RAM:27408
76 : INIT : Warm boot #55 - Restart Reason: Hardware Watchdog
78 : FS   : Mounting...
103 : FS   : Mount successful, used 75802 bytes of 957314
111 : CRC  : No program memory checksum found. Check output of crc2.py
121 : CRC  : SecuritySettings CRC   ...OK 
229 : INIT : Free RAM:24576
231 : INIT : I2C
231 : INIT : SPI not enabled
235 : INFO : Plugins: 22 [Normal] [IR_Extended] (ESP82xx Core 2_4_2, NONOS SDK 2.2.1(cfd48f3), LWIP: 2.0.3 PUYA support)
238 : WIFI : Set WiFi to STA
271 : WIFI : Connecting HTM1 attempt #0
272 : IP   : Static IP : 10.1.10.70 GW: 10.1.10.1 SN: 255.255.255.0 DNS: 10.1.10.1
274 : WIFI : Not configured in Station Mode!!: HTM1
1610 : WD   : Uptime 0 ConnectFailures 0 FreeMem 23216 WiFiStatus 0
4148 : WIFI : Connected! AP: HTM1 (00:27:22:E2:F0:52) Ch: 7 Duration: 3772 ms
4149 : IP   : Static IP : 10.1.10.70 GW: 10.1.10.1 SN: 255.255.255.0 DNS: 10.1.1
 ets Jan  8 2013,rst cause:4, boot mode:(3,7)

wdt reset
load 0x4010f000, len 1384, room 16 
tail 8
chksum 0x2d
csum 0x2d
vbb28d4a3
~ld
e⸮U64 : 

I noticed that log entry 4149 above shows DNS as 10.1.1 instead of 10.1.1.10 which is shown correct in log item 272. is that the problem?

I appreciate both.

jimmys01 commented:
Send the `reset `command thought the serial or upload the 4mb_blank.bin and try again.

TD-er commented:
It looks like some similar reboots I've seen yesterday when experimenting on ESP8266.
So I suggest for now to use DHCP addresses.
I have not looked at the related implementation of ESP32 code, but one of the things I am changing is how to tell the Arduino part of the libraries that there should be no DHCP client active.
This has (recently?) changed in the esp8266/Arduino code and was giving issues here also.

What happened here is that after N reboots it managed to continue without reboots, but sometimes N can be as high as 100 reboots.
This process of connecting to WiFi is a bit too timing critical if you ask me and it looks like slight modifications in timing in this process can make the difference of a crash during (re)connect or not.

jimmys01 committed:
[IR TX RX] Update IR library to 2.6.3

jimmys01 committed:
[IR TX RX] Update code to match new library updates

New library additions simplify the code
Added the common P016 and P035 defines in define_plugin_sets.h because of recuring defines to each other
Fixed a typo in _Plugin.ino mentioned in an Issue
Added extended AC commands ability to 252 build as well

jimmys01 commented:
[normal_core_252_IR_ESP8266_4M.zip](https://github.com/letscontrolit/ESPEasy/files/3359492/normal_core_252_IR_ESP8266_4M.zip)
If anyone wants to test it

jimmys01 commented:
I need help with that one, I do not know why it fails in builds that do not have anything to do with IR

> /home/travis/build/letscontrolit/ESPEasy/src/ESPEasy.ino:1114:17: error: 'decode_type_t' was not declared in this scope

ESPEasy.ino does not have 1114 lines and certenly does not have decode_type_t in it. This is about line 471 in P035

TD-er commented:
About the build error line numbers...
How is your Git client setup regarding line endings?
Usually in Windows it is something like "convert to CRLF on checkout, commit using Unix line endings"
If you're committing using Windows line endings, you will get such a big offset in the reported line numbers.
It is really sad to see this system is still causing all these errors, even after all those years.

And about the error itself. It looks like you're missing an include there.

happytm commented:
With your new uploaded bin file my problem with fixed ip instability gone. Now I have stable version without MPR121 & APDS9960. May be one of them was causing problem.

Thanks

jimmys01 commented:
> 
> 
> With your new uploaded bin file my problem with fixed ip instability gone. Now I have stable version without MPR121 & APDS9960.
> 
> Thanks

Hmm... this one was on 252 core version. That or some of the [testing] plugins caused your issues

happytm commented:
We can find out if using 252 core and including these 2 plugins if it stays stable then it is core issue. If not stable then it is plugins issue. I forgot MPU6050 plugin.

jimmys01 committed:
Forward declare functions

jimmys01 commented:
Ok happytm, tomorrow I will compile for you some bins.

TD-er commented:
Well, I think it is somewhat random.
I'm seeing the same here in my tests, both with core 2.5.2 and core 2.6.0 (not even released version)
One build it is working fine and the other it is failing to connect. Even when I only changed something in a line of text for a log entry.

@jimmys01 Are you building in Windows or in Linux?

jimmys01 commented:
This build is failing as well... In my PC it is compiling. Weird stuff.. goodnight

TD-er commented:
Are you building in Windows or in Linux?

jimmys01 commented:
Windows

Are you building in Windows or in Linux?

TD-er commented:
Then I think you should check your Git client settings for the line ending conversions.

It should have no effect on the build/link issues regarding WiFi stability.
I just tested a few builds here both on Windows and Linux and both seem to act the same with the same source code (and clean builds)

jimmys01 commented:
Thanks for giving me a sanity check. I hope I will find some time and fix this today, after this update I consider the IR plugins to be mature. @TD-er This PR became very busy, should I close it and open a new one?

Edit: cores 252_IR are compiling, non IR and normal IR builds are not. Anyway I found a way to make the compiler happy.

jimmys01 committed:
Avoid using enums in function declarations

happytm commented:
Jimmy,

Can you please compile  your **normal_core_252_IR_ESP8266_4M.zip**  from couple of days ago plus APDS-9960,MPR121 & MPU6050 for me whenever you get a chance.

Thanks

jimmys01 commented:
> 
> 
> Jimmy,
> 
> Can you please compile your **normal_core_252_IR_ESP8266_4M.zip** from couple of days ago plus APDS-9960,MPR121 & MPU6050 for me whenever you get a chance.
> 
> Thanks

[firmware.zip](https://github.com/letscontrolit/ESPEasy/files/3366982/firmware.zip)
Here you are, happy testing

happytm commented:
Happy testing is understatement. 
I am beyond happy.I was trying to compile for last 4 days with PlatformIO/Atom without success.Now I can go ahead with my projects and these plugins are all I will need in any situation in future specially I feel IR plugins are cheapest and simplest (because you made it simplest) input/output sensors with unlimited potential.My friend at https://github.com/enesbcs/rpieasy is trying hard to implement IR sensors for Raspberry Pi but he is finding it hard to do so.Hopefully he succeed. 

I can confirm your firmware worked flawless even with fixed IP.

I truly appreciate your effort to make IR plugins for ESPEasy what it is today.Hopefully they become part of normal compile process for ESPEasy team.I believe there is no other sensor which is so cheap(10 cents a pair),simpler in hardware design ,and smaller in size and can do so much for input/output tasks.

Thanks

jimmys01 commented:
Thank @crankyoldgit for the IR plugins, I mostly adapted his example code to work with ESP easy
